runtime/p2p.ss

```diff
@@ -0,0 +1,191 @@
+;; ~~~~~~~~ p2p.ss ~~~~~~~~~
+;;
+;; p2p communication methods for glow
+
+(export #t)
+
+;;~~~~~~~~~~~~~~~~~~~~Import~~~~~~~~~~~~~~~
+(import
+  :std/misc/channel
+  :std/iter
+  :std/net/bio
+  :std/sugar
+  :std/format
+  :std/os/pid
+  :clan/debug
+  :clan/exception
+  :gerbil/gambit/exceptions
+  :clan/path-config
+  :clan/shell
+  :clan/io
+  :gerbil/gambit/ports
+  :mukn/ethereum/known-addresses
+  ./pb/private-key
+  :std/misc/uuid
+  ../compiler/project/runtime-2
+    :gerbil/gambit/threads
+    :vyzo/libp2p
+    :vyzo/libp2p/client
+  :vyzo/libp2p/daemon)
+
+
+
+
+;;~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~Direct p2p communication methods
+;;
+;;
+;; peer-multiaddr-str: destination multiaddress,
+;; It has the additional constraint that it needs to contain a peerId,
+;; so we can verify the recipient's identity.
+;;
+;; host-addresses: Multi addresses this participant listens to on their host machine.
+;;
+;; contents: string
+;;
+;; This is used to connect to a destination address of another participant,
+;; and send the contents over the opened connection.
+;; If the other participant is not online,
+;; it will poll until `timeout'.
+(def (dial-and-send-contents libp2p-client dest-address-str contents timeout: (timeout 10))
+  (let* ((self (libp2p-identify libp2p-client))
+         (peer-multiaddr (string->peer-info dest-address-str)))
+    (for (p (peer-info->string* self))
+      (displayln "I am " p))
+    (displayln "Connecting to " dest-address-str)
+    (libp2p-connect/poll libp2p-client peer-multiaddr timeout: timeout)
+    (let (s (libp2p-stream libp2p-client peer-multiaddr [chat-proto]))
+      (chat-writer s contents)
+      (stream-close s))))
+
+;; This is used to connect to a destination address of another participant,
+;; and send the contents over the opened connection.
+;; If the other participant is not online,
+;; it will poll until `timeout'.
+(def (libp2p-connect/poll libp2p-client peer-multiaddr timeout: (timeout #f))
+  (try (libp2p-connect libp2p-client peer-multiaddr)
+    (catch (e)
+      (printf "Unable to connect to client ~a ...\n" (peer-info->string peer-multiaddr))
+      (display-exception e)
+      (if (and timeout (> timeout 0))
+          (let ()
+            (displayln "Polling again in 1s...")
+            (thread-sleep! 1)
+            (libp2p-connect/poll libp2p-client peer-multiaddr timeout: (- timeout 1)))
+          (error "Timeout while trying to connect to client.")))))
```

Comment:
When I was doing a similar thing, polling every second, in my `tcp` branch, @fare said
> Also, a 1 second delay is probably not appropriate for anything but local network testing. E.g. to avoid congestion on the Internet, retry protocols usually need to use exponential backoff. See the `retry` function in `clan/concurrency`.

If that's what we want, this could probably be re-written to use exponential backoff with a pattern like:
```scheme
(retry
  retry-window: retry-window
  max-window: max-window
  max-retries: max-retries
  (lambda ()
    (libp2p-connect libp2p-client peer-multiaddr)))
```

Reply:
Good idea! I will fix this.